### PR TITLE
Improve Object3D `add` and `remove`

### PR DESCRIPTION
**Pull request by brianchirls**
- Dispatch "add" and "remove" events on parents in addition to child so new child objects can be monitored
- Moved "added" and "removed" dispatch after children array is modified so the operation is complete before the event fires
- Added unit test for all the above

### CODE REVIEW

File: src/core/Object3D.js [hunk 1]
```diff
@@ -332,10 +332,11 @@ Object.assign( Object3D.prototype, EventDispatcher.prototype, {
 			}
 
 			object.parent = this;
-			object.dispatchEvent( { type: 'added' } );
-
 			this.children.push( object );
 
+			object.dispatchEvent( { type: 'added' } );
+			this.dispatchEvent( { type: 'add', child: object } );
```
Comment (mrdoob) [inline src/core/Object3D.js]: `childAdded` and `childRemoved` may be better names?

Comment (brianchirls) [inline src/core/Object3D.js]: Okay, changed.

**Comment by brianchirls** [comment]
@mrdoob updated per your comment

**Comment by brianchirls** [comment]
Bump for approval, @mrdoob. I made the changes you requested.

**Comment by Fyrestar** [comment]
Is creating an object for the event really necessary? It coult be a reused one.

**Comment by mrdoob** [comment]
I guess the problem with this one was that it has two changes in one. Dispatching the event after the operation is done makes sense. The other change it would be better in a different PR.

**Pull request closed without merging**